k9mail/src/main/java/security/planck/messaging/MessagingRepository.kt

```diff
@@ -0,0 +1,269 @@
+import com.fsck.k9.extensions.hasToBeDecrypted
+import com.fsck.k9.extensions.isMessageIncomplete
+import com.fsck.k9.mail.Flag
+import com.fsck.k9.mail.internet.MimeHeader
+import com.fsck.k9.mail.internet.MimeMessage
+import com.fsck.k9.mailstore.LocalMessage
+import com.fsck.k9.mailstore.MessageViewInfo
+import com.fsck.k9.mailstore.MessageViewInfoExtractor
+import com.fsck.k9.planck.DispatcherProvider
+import com.fsck.k9.planck.PlanckProvider
+import com.fsck.k9.planck.PlanckUtils
+import com.fsck.k9.planck.infrastructure.exceptions.KeyMissingException
+import com.fsck.k9.planck.infrastructure.extensions.flatMapSuspend
+import com.fsck.k9.ui.messageview.MessageViewEffect
+import com.fsck.k9.ui.messageview.MessageViewState
+import com.fsck.k9.ui.messageview.MessageViewUpdate
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class MessagingRepository @Inject constructor(
+    private val controller: MessagingController,
+    private val planckProvider: PlanckProvider,
+    private val infoExtractor: MessageViewInfoExtractor,
+    private val appIoScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
+) {
+
+    suspend fun loadMessage(
+        messageViewUpdate: MessageViewUpdate,
+    ) = withContext(dispatcherProvider.io()) {
+        messageViewUpdate.stateFlow.value = MessageViewState.Loading
+        messageViewUpdate.effectFlow.value = MessageViewEffect.NoEffect
+        loadMessageFromDatabase(
+            messageViewUpdate
+        ).join()
+    }
+
+    private fun loadMessageFromDatabase(
+        messageViewUpdate: MessageViewUpdate,
+    ) = appIoScope.launch {
+        with(messageViewUpdate) {
+            kotlin.runCatching {
+                controller.loadMessage(account, messageReference.folderName, messageReference.uid)
+            }.onFailure {
+                Timber.e(it)
+                effectFlow.value = MessageViewEffect.ErrorLoadingMessage(it)
+            }.onSuccess { message ->
+                message?.let {
+                    messageLoaded(message, messageViewUpdate)
+                } ?: let { effectFlow.value = MessageViewEffect.ErrorLoadingMessage() }
+            }
+        }
+    }
+
+    private suspend fun messageLoaded(
+        message: LocalMessage,
+        messageViewUpdate: MessageViewUpdate,
+    ) {
+        message.recoverRating()
+        val loadedState = if (!message.hasToBeDecrypted()) {
+            MessageViewState.DecryptedMessageLoaded(message)
+        } else {
+            MessageViewState.EncryptedMessageLoaded(message)
+        }
+
+        with(messageViewUpdate) {
+            withContext(dispatcherProvider.main()) {
+                stateFlow.value = loadedState
+            }
+            if (message.isMessageIncomplete()) {
```

Comment:
if as message is both incomplete and to be decrypted how does this work? You first enter this if, download it, then run through this message loaded function again and go to the decrypt?

Reply:
Yep, a message incomplete by definition is not a message to be decrypted so no problem. (Message is required to be fully dowloaded, etc to be elegible for decrypt). Actually it may be interesting to re-check behavior with SMime messages here.